---
title: "Allen_2025a_Analysis__Figure_4_pseudotime_summarize_markers"
description: description
author: "Aaron M. Allen"
date: 'Last update: `r date()`'
output:
  html_document:
    number_sections: true
    code_folding: show
    code_download: true
    theme: cerulean
    df_print: paged
    fig_width: 8.5
    fig_height: 5
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: false
    toc_depth: 6
---


```{css, echo = FALSE}
pre, code {white-space:pre !important; overflow-x:auto}
.tocify-item {white-space:pre}
```



```{r echo = FALSE, cache = FALSE}

```


```{r setup, echo = FALSE, cache = FALSE}
ggplot2::theme_set(ggplot2::theme_grey())
ggplot2::theme_update(panel.background = ggplot2::element_rect(fill = "transparent", colour = NA),
                      plot.background = ggplot2::element_rect(fill = "transparent", colour = NA))

rmd_name <- gsub(pattern = ".Rmd", replacement = "", x = knitr::current_input(), ignore.case = TRUE)
knitr::opts_chunk$set(dev = c("png", "cairo_pdf"),
                      dev.args=list(bg="transparent"),
                      fig.align = "center",
                      fig.height = 5,
                      fig.width = 8.5,
                      dpi = 300,
                      pdf.options(encoding = "ISOLatin9.enc"),
                      fig.path = paste0("../analyses/figures/", rmd_name, "-",
                                        format(Sys.time(), "%Y%m%d_%H%M%S"), "/"),
                      fig.retina = 1,
                      warning = TRUE,
                      message = TRUE)
```











# Functions

```{r}
library(tidyverse)
library(zoo)

reduce_columns_to_100 <- function(df) {
    num_cols <- ncol(df)
    if (num_cols <= 100) {
        return(df)
    }
    col_indices <- seq(1, num_cols, length.out = 100)
    df_reduced <- df %>%
        rowwise() %>%
        mutate(across(
            everything(),
            ~ zoo::rollmean(.x, k = ceiling(num_cols / 100), fill = NA, align = "right")[col_indices]
            )) %>%
        ungroup() %>%
        select(1:100)
    return(df_reduced)
}

```


```{r}
library(dplyr)
library(purrr)

average_reduce_columns <- function(df, target_cols = 100) {
	df <- df %>%
		as.data.frame() %>% 
		column_to_rownames("gene")
	n_cols <- ncol(df)
	row_names <- rownames(df)
	group_size <- ceiling(n_cols / target_cols)
	df_grouped <- df %>%
		as.data.frame() %>%
		split.default(ceiling(seq_along(df) / group_size)) %>%
		map_dfc(~ rowMeans(.x, na.rm = TRUE))
	colnames(df_grouped) <- paste0("V", seq_len(ncol(df_grouped)))
	rownames(df_grouped) <- row_names
	return(df_grouped)
}

reduce_dataframes <- function(df_list) {
	df_list %>% map(~ average_reduce_columns(.x))
}
```












# Spot check

```{r}
test_smooth_spline <- read_csv("../analyses/markers/subclustering/RNA_snn_res_10/pseudotime_v2/RNA_snn_res.1/subcluster_dims_10/subclustering_RNA_snn_res_10__cluster_027__pseudotime_smooth_spline_markers.csv")
test_smooth_spline
```





```{r}
small_test_smooth_spline <- test_smooth_spline %>% 
	filter(gene %in% c("Imp","br", "pdm3", "bab2", "dati")) 
small_test_smooth_spline
```



```{r}
average_reduce_columns(small_test_smooth_spline) %>% 
	bind_cols(gene = c("pdm3", "dati", "bab2", "br")) %>% 
	gather("pseudotime", "smoothed_exp", -gene) %>% 
	mutate(pseudotime = str_remove(pseudotime, "V") %>% as.numeric()) %>% 
	# filter(gene == "br") %>% 
	ggplot(aes(x = pseudotime, y = smoothed_exp, colour = gene)) +
		geom_line()
```












# Read in all data


```{r}
clusters_keep <- c(
	10,16,17,18,19,20,22,23,24,26,27,29,30,33,36,38,39,42,43,46,
	48,49,50,51,53,55,56,57,58,62,63,64,66,69,71,72,73,74,75,81,
	82,84,85,86,87,88,89,90,94,95,98,100,101,105,106,107,109,110,
	113,115,117,121,128,129,132,134,135,138,145,146,148,150,151,
	152,159,161,166,169,171,175,180,182,193,195
)
clusters_keep
```


```{r}
list.files(path = "../analyses/markers/subclustering/RNA_snn_res_10/pseudotime_v2/RNA_snn_res.1/subcluster_dims_10/")
```



```{r}
smooth_spline_files <- list.files(path = "../analyses/markers/subclustering/RNA_snn_res_10/pseudotime_v2/RNA_snn_res.1/subcluster_dims_10/", pattern = "pseudotime_smooth_spline_markers", full.names = TRUE)
smooth_spline_files
```

```{r}
smooth_spline_files[2] %>% 
	str_remove("../analyses/markers/subclustering/RNA_snn_res_10/pseudotime_v2/RNA_snn_res.1/subcluster_dims_10//subclustering_RNA_snn_res_10__") %>% 
	str_remove("__pseudotime_smooth_spline_markers.csv")
```

```{r}
paste0("_cluster_", sprintf("%03d", clusters_keep))
```


```{r}
smooth_spline_files <- data.frame(file_names = smooth_spline_files,
		   cluster_number = smooth_spline_files %>% 
								str_remove("../analyses/markers/subclustering/RNA_snn_res_10/pseudotime_v2/RNA_snn_res.1/subcluster_dims_10//subclustering_RNA_snn_res_10__") %>% 
								str_remove("__pseudotime_smooth_spline_markers.csv")
	) %>% 
	filter(cluster_number %in% paste0("cluster_", sprintf("%03d", clusters_keep))) %>% 
	pull(file_names)
smooth_spline_files
```












```{r message=FALSE, warning=FALSE}
smooth_spline_list <- list()
for (i in seq_along(smooth_spline_files)) {
	message(paste0("Loading list .. ", i, " .. of ", length(smooth_spline_files)))
	smooth_spline_list[[i]] <- read_csv(smooth_spline_files[[i]])#, show_col_types = FALSE)
}
```

```{r}
length(smooth_spline_list)
```


```{r}
smooth_spline_list_names <- list()
for (i in seq_along(smooth_spline_files)) {
	smooth_spline_list_names[[i]] <- smooth_spline_files[i] %>% 
										str_remove("../analyses/markers/subclustering/RNA_snn_res_10/pseudotime_v2/RNA_snn_res.1/subcluster_dims_10//subclustering_RNA_snn_res_10__") %>% 
										str_remove("__pseudotime_smooth_spline_markers.csv")
}
names(smooth_spline_list) <- smooth_spline_list_names
```

```{r}
length(smooth_spline_list_names)
```




```{r}
smooth_spline_list[[1]]
```


```{r}
reduced_smooth_spline_list <- reduce_dataframes(smooth_spline_list)
```

```{r}
reduced_smooth_spline_list[[1]]
```

```{r}
rownames(reduced_smooth_spline_list[[1]])[1:5]
```

```{r}
names(reduced_smooth_spline_list)[[1]]
```




## Check individual genes


### Imp

```{r}
reduced_imp <- list()
for (i in seq_along(reduced_smooth_spline_list)) {
	reduced_imp[[i]] <- reduced_smooth_spline_list[[i]] %>% 
		filter(row.names(reduced_smooth_spline_list[[i]]) %in% c("Imp")) %>% 
		mutate(cluster = names(reduced_smooth_spline_list)[[i]]) %>% 
		dplyr::relocate(cluster)
}
bind_rows(reduced_imp)
```



```{r}
bind_rows(reduced_imp) %>% 
	# rownames_to_column("cluster") %>% 
	gather("pseudotime", "smoothed_exp", -cluster) %>% 
	mutate(pseudotime = str_remove(pseudotime, "V") %>% as.numeric()) %>% 
	ggplot(aes(x = pseudotime, y = smoothed_exp, fill = cluster)) +
		geom_line(alpha = 0.1)
```


```{r}
bind_rows(reduced_imp) %>% 
	# rownames_to_column("cluster") %>% 
	gather("pseudotime", "smoothed_exp", -cluster) %>% 
	group_by(pseudotime) %>% 
	summarise(mean_exp = mean(smoothed_exp),
			  conf_int = 0.95*(sd(smoothed_exp)/sqrt(n()))) %>% 
	mutate(pseudotime = str_remove(pseudotime, "V") %>% as.numeric()) %>% 
	ggplot(aes(x = pseudotime, y = mean_exp)) +
		geom_line() +
		geom_ribbon(aes(ymin = mean_exp - conf_int, ymax = mean_exp + conf_int), 
					fill = "lightsteelblue2", alpha = 0.3) +
		theme_classic()
```



### br

```{r}
reduced_br <- list()
for (i in seq_along(reduced_smooth_spline_list)) {
	reduced_br[[i]] <- reduced_smooth_spline_list[[i]] %>% 
		filter(row.names(reduced_smooth_spline_list[[i]]) %in% c("br")) %>% 
		mutate(cluster = names(reduced_smooth_spline_list)[[i]]) %>% 
		dplyr::relocate(cluster)
}
bind_rows(reduced_br)
```


```{r}
bind_rows(reduced_br) %>% 
	gather("pseudotime", "smoothed_exp", -cluster) %>% 
	mutate(pseudotime = str_remove(pseudotime, "V") %>% as.numeric()) %>% 
	ggplot(aes(x = pseudotime, y = smoothed_exp, fill = cluster)) +
		geom_line(alpha = 0.1)
```


```{r}
bind_rows(reduced_br) %>% 
	gather("pseudotime", "smoothed_exp", -cluster) %>% 
	group_by(pseudotime) %>% 
	summarise(mean_exp = mean(smoothed_exp),
			  conf_int = 0.95*(sd(smoothed_exp)/sqrt(n()))) %>% 
	mutate(pseudotime = str_remove(pseudotime, "V") %>% as.numeric()) %>% 
	ggplot(aes(x = pseudotime, y = mean_exp)) +
		geom_line() +
		geom_ribbon(aes(ymin = mean_exp - conf_int, ymax = mean_exp + conf_int), 
					fill = "lightsteelblue2", alpha = 0.3) +
		theme_classic()
```



### dati

```{r}
reduced_dati <- list()
for (i in seq_along(reduced_smooth_spline_list)) {
	reduced_dati[[i]] <- reduced_smooth_spline_list[[i]] %>% 
		filter(row.names(reduced_smooth_spline_list[[i]]) %in% c("dati")) %>% 
		mutate(cluster = names(reduced_smooth_spline_list)[[i]]) %>% 
		dplyr::relocate(cluster)
}
bind_rows(reduced_dati)
```


```{r}
bind_rows(reduced_dati) %>% 
	gather("pseudotime", "smoothed_exp", -cluster) %>% 
	mutate(pseudotime = str_remove(pseudotime, "V") %>% as.numeric()) %>% 
	ggplot(aes(x = pseudotime, y = smoothed_exp, fill = cluster)) +
		geom_line(alpha = 0.1)
```


```{r}
bind_rows(reduced_dati) %>% 
	gather("pseudotime", "smoothed_exp", -cluster) %>% 
	group_by(pseudotime) %>% 
	summarise(mean_exp = mean(smoothed_exp),
			  conf_int = 0.95*(sd(smoothed_exp)/sqrt(n()))) %>% 
	mutate(pseudotime = str_remove(pseudotime, "V") %>% as.numeric()) %>% 
	ggplot(aes(x = pseudotime, y = mean_exp)) +
		geom_line() +
		geom_ribbon(aes(ymin = mean_exp - conf_int, ymax = mean_exp + conf_int), 
					fill = "lightsteelblue2", alpha = 0.3) +
		theme_classic()
```








## TFs

### r6.30


```{r load-tfs}
r630 <- read_tsv("../../../aallen/genomes/dmel/r6.30/gtf/dmel-all-r6.30.gtf.gz",
                 col_names = c("sequence","source","feature","start","end","score","strand","phase","attributes")) %>%
                 separate(attributes,c("gene_id","gene_symbol","transcript_id","transcript_symbol"),sep = ";")
# r630

r630_clean <- r630 %>% 
    filter(feature == "gene") %>% 
    select(gene_id, gene_symbol) %>% 
    mutate(gene_id = gene_id %>% 
                       str_remove_all(pattern = '"') %>% 
                       str_remove(pattern = 'gene_id '),
           gene_symbol = gene_symbol %>% 
                           str_remove_all(pattern = '"') %>% 
                           str_remove(pattern = 'gene_symbol ')
           )
# r630_clean

tfs_fbgn <- read_csv("../../../aallen/gene_lists/gene_sets/FlyBase_IDs_allTFs.txt")
# tfs_fbgn

tfs_symbol <- r630_clean %>% 
    filter(gene_id %in% tfs_fbgn$FBgn) %>% 
    select(gene_symbol) %>% 
    mutate(gene_symbol = gene_symbol %>% 
                           str_remove(pattern = ' ')
           )
# tfs_symbol
```


```{r}
tfs_symbol$gene_symbol
```

```{r}
names(reduced_smooth_spline_list)[[i]]
```




```{r}
reduced_tfs <- list()
for (i in seq_along(reduced_smooth_spline_list)) {
	reduced_tfs[[i]] <- reduced_smooth_spline_list[[i]] %>% 
		rownames_to_column("gene") %>% 
		filter(gene %in% tfs_symbol$gene_symbol) %>% 
		mutate(subcluster = paste0("subcluster_", i)) %>% 
		select(subcluster, gene)
}
reduced_tfs <- bind_rows(reduced_tfs)
reduced_tfs
```


```{r}
reduced_tfs %>% 
	count(gene) %>% 
	arrange(desc(n))
```










# Repeated TFs

```{r}
reduced_tfs %>% 
	count(gene) %>% 
	arrange(desc(n))
```



```{r}
repeated_tfs <- reduced_tfs %>% 
	count(gene) %>% 
	arrange(desc(n)) %>% 
	filter(n > 15) %>% 
	pull(gene) %>% 
	sort()
repeated_tfs
```


```{r warning=FALSE, message=FALSE}

genes <- repeated_tfs
for (ii in seq_along(genes)) {
	
	reduced_goi <- list()
	for (i in seq_along(reduced_smooth_spline_list)) {
		reduced_goi[[i]] <- reduced_smooth_spline_list[[i]] %>% filter(row.names(reduced_smooth_spline_list[[i]]) %in% c(genes[[ii]]))
	}
	
	p1 <- bind_rows(reduced_goi) %>% 
		rownames_to_column("cluster") %>% 
		gather("pseudotime", "smoothed_exp", -cluster) %>% 
		mutate(pseudotime = str_remove(pseudotime, "V") %>% as.numeric()) %>% 
		ggplot(aes(x = pseudotime, y = smoothed_exp, fill = cluster)) +
			geom_line(alpha = 0.1) +
			ggtitle(genes[[ii]])
	plot(p1)
	
	p1 <- bind_rows(reduced_goi) %>% 
		rownames_to_column("cluster") %>% 
		gather("pseudotime", "smoothed_exp", -cluster) %>% 
		group_by(pseudotime) %>% 
		summarise(mean_exp = mean(smoothed_exp)) %>% 
		mutate(pseudotime = str_remove(pseudotime, "V") %>% as.numeric()) %>% 
		ggplot(aes(x = pseudotime, y = mean_exp)) +
			geom_line() +
			ggtitle(genes[[ii]])
	plot(p1)
	
}


```



```{r warning=FALSE, message=FALSE}

genes <- c("dan", "danr")
for (ii in seq_along(genes)) {
	
	reduced_goi <- list()
	for (i in seq_along(reduced_smooth_spline_list)) {
		reduced_goi[[i]] <- reduced_smooth_spline_list[[i]] %>% filter(row.names(reduced_smooth_spline_list[[i]]) %in% c(genes[[ii]]))
	}
	
	p1 <- bind_rows(reduced_goi) %>% 
		rownames_to_column("cluster") %>% 
		gather("pseudotime", "smoothed_exp", -cluster) %>% 
		mutate(pseudotime = str_remove(pseudotime, "V") %>% as.numeric()) %>% 
		ggplot(aes(x = pseudotime, y = smoothed_exp, fill = cluster)) +
			geom_line(alpha = 0.1) +
			ggtitle(genes[[ii]])
	plot(p1)
	
	p1 <- bind_rows(reduced_goi) %>% 
		rownames_to_column("cluster") %>% 
		gather("pseudotime", "smoothed_exp", -cluster) %>% 
		group_by(pseudotime) %>% 
		summarise(mean_exp = mean(smoothed_exp)) %>% 
		mutate(pseudotime = str_remove(pseudotime, "V") %>% as.numeric()) %>% 
		ggplot(aes(x = pseudotime, y = mean_exp)) +
			geom_line() +
			ggtitle(genes[[ii]])
	plot(p1)
	
}


```



```{r}
reduced_all <- list()
for (i in seq_along(reduced_smooth_spline_list)) {
	reduced_all[[i]] <- reduced_smooth_spline_list[[i]] %>% 
		rownames_to_column("gene") %>% 
		mutate(subcluster = paste0("subcluster_", i)) %>% 
		dplyr::relocate(subcluster, gene)
}
reduced_all <- bind_rows(reduced_all)
reduced_all
```

```{r}
mean_reduced_tfs <- reduced_all %>% 
	filter(gene %in% repeated_tfs) %>% 
	gather("pseudotime", "smoothed_exp", -subcluster, -gene) %>% 
	mutate(pseudotime = str_remove(pseudotime, "V") %>% as.numeric()) %>% 
	group_by(gene, pseudotime) %>% 
	summarise(mean_exp = mean(smoothed_exp)) %>% 
	spread(pseudotime, mean_exp) %>% 
	column_to_rownames("gene") %>% 
	select(1:90)
mean_reduced_tfs
```


```{r}
library(RColorBrewer)
library(ComplexHeatmap)
```


```{r}
#K means with 6 groups
htkm <- Heatmap(
  mean_reduced_tfs,
  name                         = "z-score",
  col                          = circlize::colorRamp2(seq(from = -0.5, to = 0.5,length = 11),rev(brewer.pal(11, "Spectral"))),
  show_row_names               = TRUE,
  show_column_names            = FALSE,
  row_names_gp                 = gpar(fontsize = 12),
  km                           = 10,
  row_title_rot                = 0,
  cluster_rows                 = TRUE,
  cluster_row_slices           = FALSE,
  cluster_columns              = FALSE,
  use_raster = FALSE)

#Ward.D2 Hierarchical Clustesubclustered_cluster
hthc <- Heatmap(
  mean_reduced_tfs,
  name                         = "z-score",
  col                          = circlize::colorRamp2(seq(from = -0.5, to = 0.5, length = 11),rev(brewer.pal(11, "Spectral"))),
  show_row_names               = TRUE,
  show_column_names            = FALSE,
  row_names_gp                 = gpar(fontsize = 12),
  clustering_method_rows       = "ward.D2",
  clustering_method_columns    = "ward.D2",
  row_title_rot                = 0,
  cluster_rows                 = TRUE,
  cluster_row_slices           = FALSE,
  cluster_columns              = FALSE,
  use_raster = FALSE)
```



```{r fig.width=8, fig.height=4}
print(htkm)
print(hthc)
```











```{r}
mean_reduced_tfs_scaled <- t(apply(mean_reduced_tfs, 1, scale))
mean_reduced_tfs_scaled <- as.data.frame(mean_reduced_tfs_scaled)
```


```{r}
#K means with 6 groups
htkm <- Heatmap(
  mean_reduced_tfs_scaled,
  name                         = "z-score",
  col                          = circlize::colorRamp2(seq(from = -1, to = 1,length = 11),rev(brewer.pal(11, "Spectral"))),
  show_row_names               = TRUE,
  show_column_names            = FALSE,
  row_names_gp                 = gpar(fontsize = 12),
  km                           = 10,
  row_title_rot                = 0,
  cluster_rows                 = TRUE,
  cluster_row_slices           = FALSE,
  cluster_columns              = FALSE,
  use_raster = TRUE)

#Ward.D2 Hierarchical Clustesubclustered_cluster
hthc <- Heatmap(
  mean_reduced_tfs_scaled,
  name                         = "z-score",
  col                          = circlize::colorRamp2(seq(from = -1, to = 1, length = 11),rev(brewer.pal(11, "Spectral"))),
  show_row_names               = TRUE,
  show_column_names            = FALSE,
  row_names_gp                 = gpar(fontsize = 12),
  clustering_method_rows       = "ward.D2",
  clustering_method_columns    = "ward.D2",
  row_title_rot                = 0,
  cluster_rows                 = TRUE,
  cluster_row_slices           = FALSE,
  cluster_columns              = FALSE,
  use_raster = TRUE)
```



```{r fig.width=8, fig.height=4}
print(htkm)
print(hthc)
```


```{r warning=FALSE, message=FALSE, fig.width=8, fig.height=4}

cluster_methods <- c("ward.D","ward.D2","single","complete", "average", "mcquitty", "median", "centroid")
for (i in seq_along(cluster_methods)) {
	hthc <- Heatmap(
		mean_reduced_tfs_scaled,
		name                         = "z-score",
		col                          = circlize::colorRamp2(seq(from = -1, to = 1, length = 11),rev(brewer.pal(11, "Spectral"))),
		show_row_names               = TRUE,
		show_column_names            = FALSE,
		row_names_gp                 = gpar(fontsize = 12),
		clustering_method_rows       = cluster_methods[[i]],
		clustering_method_columns    = cluster_methods[[i]],
		row_title_rot                = 0,
		cluster_rows                 = TRUE,
		cluster_row_slices           = FALSE,
		cluster_columns              = FALSE,
		use_raster = TRUE)
	print(hthc)	
}
```






```{r}
hand_picked_genes <- c("Imp", "mamo", "dan","danr","pdm3","pdm2","RunxA","bab1","bab2","br","Eip93F","ab","sr","jim","dati","pros","zld","lov","fru")
```


```{r}
mean_reduced_hand_picked <- reduced_all %>% 
	filter(gene %in% hand_picked_genes) %>% 
	gather("pseudotime", "smoothed_exp", -subcluster, -gene) %>% 
	mutate(pseudotime = str_remove(pseudotime, "V") %>% as.numeric()) %>% 
	group_by(gene, pseudotime) %>% 
	summarise(mean_exp = mean(smoothed_exp)) %>% 
	spread(pseudotime, mean_exp) %>% 
	column_to_rownames("gene") %>% 
	select(1:90)
mean_reduced_hand_picked
```




```{r}
mean_reduced_hand_picked_scaled <- t(apply(mean_reduced_hand_picked, 1, scale))
mean_reduced_hand_picked_scaled <- as.data.frame(mean_reduced_hand_picked_scaled)
```


```{r}
#K means with 6 groups
htkm <- Heatmap(
  mean_reduced_hand_picked_scaled,
  name                         = "z-score",
  col                          = circlize::colorRamp2(seq(from = -1, to = 1,length = 11),rev(brewer.pal(11, "Spectral"))),
  show_row_names               = TRUE,
  show_column_names            = FALSE,
  row_names_gp                 = gpar(fontsize = 12),
  km                           = 10,
  row_title_rot                = 0,
  cluster_rows                 = TRUE,
  cluster_row_slices           = FALSE,
  cluster_columns              = FALSE,
  use_raster = TRUE)

#Ward.D2 Hierarchical Clustesubclustered_cluster
hthc <- Heatmap(
  mean_reduced_hand_picked_scaled,
  name                         = "z-score",
  col                          = circlize::colorRamp2(seq(from = -1, to = 1, length = 11),rev(brewer.pal(11, "Spectral"))),
  show_row_names               = TRUE,
  show_column_names            = FALSE,
  row_names_gp                 = gpar(fontsize = 12),
  clustering_method_rows       = "ward.D2",
  clustering_method_columns    = "ward.D2",
  row_title_rot                = 0,
  cluster_rows                 = TRUE,
  cluster_row_slices           = FALSE,
  cluster_columns              = FALSE,
  use_raster = TRUE)
```



```{r fig.width=8, fig.height=4}
print(htkm)
print(hthc)
```


```{r}

Heatmap(
  mean_reduced_hand_picked_scaled,
  name                         = "z-score",
  col                          = circlize::colorRamp2(seq(from = -1, to = 1, length = 11),rev(brewer.pal(11, "Spectral"))),
  show_row_names               = TRUE,
  show_column_names            = FALSE,
  row_names_gp                 = gpar(fontsize = 12),
  clustering_method_rows       = "ward.D2",
  clustering_method_columns    = "ward.D2",
  row_title_rot                = 0,
  cluster_rows                 = TRUE,
  cluster_row_slices           = FALSE,
  cluster_columns              = FALSE,
  use_raster = FALSE)

```




```{r warning=FALSE, message=FALSE, fig.width=8, fig.height=4}

cluster_methods <- c("ward.D","ward.D2","single","complete", "average", "mcquitty", "median", "centroid")
for (i in seq_along(cluster_methods)) {
	hthc <- Heatmap(
		mean_reduced_hand_picked_scaled,
		name                         = "z-score",
		col                          = circlize::colorRamp2(seq(from = -1, to = 1, length = 11),rev(brewer.pal(11, "Spectral"))),
		show_row_names               = TRUE,
		show_column_names            = FALSE,
		row_names_gp                 = gpar(fontsize = 12),
		clustering_method_rows       = cluster_methods[[i]],
		clustering_method_columns    = cluster_methods[[i]],
		row_title_rot                = 0,
		cluster_rows                 = TRUE,
		cluster_row_slices           = FALSE,
		cluster_columns              = FALSE,
		use_raster = TRUE)
	print(hthc)	
}
```




```{r fig.height=12, fig.width=20}

genes <- hand_picked_genes
plist <- FeaturePlot(object = meta_cb, 
           reduction = "tsne_240pcs", 
           features = genes,
           combine = FALSE,
           ncol = 1,
           pt.size = 0.4, 
           min.cutoff = 1,
		   max.cutoff = 2,
		   # slot = "counts",
           cols = c("lightsteelblue2", "black"), 
           raster = FALSE,
           raster.dpi = c(1024, 1024),
           order = TRUE,
           coord.fixed = TRUE)

for (i in seq_along(plist)) {
	plist[[i]]$layers[[1]]$aes_params$alpha = .1
	plot(plist[[i]])
}

```













```{r}
repeated_tfs <- reduced_tfs %>% 
	count(gene) %>% 
	arrange(desc(n)) %>% 
	filter(n < 30 & n > 10) %>% 
	pull(gene) %>% 
	sort()
repeated_tfs
```

```{r warning=FALSE, message=FALSE}

genes <- repeated_tfs
for (ii in seq_along(genes)) {
	
	reduced_goi <- list()
	for (i in seq_along(reduced_smooth_spline_list)) {
		reduced_goi[[i]] <- reduced_smooth_spline_list[[i]] %>% filter(row.names(reduced_smooth_spline_list[[i]]) %in% c(genes[[ii]]))
	}
	
	p1 <- bind_rows(reduced_goi) %>% 
		rownames_to_column("cluster") %>% 
		gather("pseudotime", "smoothed_exp", -cluster) %>% 
		mutate(pseudotime = str_remove(pseudotime, "V") %>% as.numeric()) %>% 
		ggplot(aes(x = pseudotime, y = smoothed_exp, fill = cluster)) +
			geom_line(alpha = 0.1) +
			ggtitle(genes[[ii]])
	plot(p1)
	
	p1 <- bind_rows(reduced_goi) %>% 
		rownames_to_column("cluster") %>% 
		gather("pseudotime", "smoothed_exp", -cluster) %>% 
		group_by(pseudotime) %>% 
		summarise(mean_exp = mean(smoothed_exp)) %>% 
		mutate(pseudotime = str_remove(pseudotime, "V") %>% as.numeric()) %>% 
		ggplot(aes(x = pseudotime, y = mean_exp)) +
			geom_line() +
			ggtitle(genes[[ii]])
	plot(p1)
	
}


```








```{r}
#K means with 6 groups
htkm <- Heatmap(
  tf.pt.matrix,
  name                         = "z-score",
  col                          = circlize::colorRamp2(seq(from = -2, to = 2,length = 11),rev(brewer.pal(11, "Spectral"))),
  show_row_names               = TRUE,
  show_column_names            = FALSE,
  row_names_gp                 = gpar(fontsize = 12),
  km                           = 10,
  row_title_rot                = 0,
  cluster_rows                 = TRUE,
  cluster_row_slices           = FALSE,
  cluster_columns              = FALSE,
  use_raster = TRUE)

#Ward.D2 Hierarchical Clustesubclustered_cluster
hthc <- Heatmap(
  tf.pt.matrix,
  name                         = "z-score",
  col                          = circlize::colorRamp2(seq(from = -2, to = 2, length = 11),rev(brewer.pal(11, "Spectral"))),
  show_row_names               = TRUE,
  show_column_names            = FALSE,
  row_names_gp                 = gpar(fontsize = 12),
  clustering_method_rows       = "ward.D2",
  clustering_method_columns    = "ward.D2",
  row_title_rot                = 0,
  cluster_rows                 = TRUE,
  cluster_row_slices           = FALSE,
  cluster_columns              = FALSE,
  use_raster = TRUE)
```



```{r fig.width=8, fig.height=4}
print(htkm)
print(hthc)
```





# Re-spline and non-filtered sig


## Read in all data


```{r}
clusters_keep <- c(
	10,16,17,18,19,20,22,23,24,26,27,29,30,33,36,38,39,42,43,46,
	48,49,50,51,53,55,56,57,58,62,63,64,66,69,71,72,73,74,75,81,
	82,84,85,86,87,88,89,90,94,95,98,100,101,105,106,107,109,110,
	113,115,117,121,128,129,132,134,135,138,145,146,148,150,151,
	152,159,161,166,169,171,175,180,182,193,195
)
clusters_keep
```





```{r}
pseudotime_markers_files <- list.files(path = "../analyses/markers/subclustering/RNA_snn_res_10/pseudotime_v2/RNA_snn_res.1/subcluster_dims_10/", pattern = "__pseudotime_markers.csv", full.names = TRUE)
pseudotime_markers_files
```

```{r}
pseudotime_markers_files[2] %>% 
	str_remove("../analyses/markers/subclustering/RNA_snn_res_10/pseudotime_v2/RNA_snn_res.1/subcluster_dims_10//subclustering_RNA_snn_res_10__") %>% 
	str_remove("__pseudotime_markers.csv")
```

```{r}
paste0("_cluster_", sprintf("%03d", clusters_keep))
```


```{r}
pseudotime_markers_files <- data.frame(file_names = pseudotime_markers_files,
		   cluster_number = pseudotime_markers_files %>% 
								str_remove("../analyses/markers/subclustering/RNA_snn_res_10/pseudotime_v2/RNA_snn_res.1/subcluster_dims_10//subclustering_RNA_snn_res_10__") %>% 
								str_remove("__pseudotime_markers.csv")
	) %>% 
	filter(cluster_number %in% paste0("cluster_", sprintf("%03d", clusters_keep))) %>% 
	pull(file_names)
pseudotime_markers_files
```




```{r message=FALSE, warning=FALSE}
pseudotime_markers_list <- list()
for (i in seq_along(pseudotime_markers_files)) {
	message(paste0("Loading list .. ", i, " .. of ", length(pseudotime_markers_files)))
	pseudotime_markers_list[[i]] <- read_csv(pseudotime_markers_files[[i]])#, show_col_types = FALSE)
}
```

```{r}
length(pseudotime_markers_list)
```


```{r}
pseudotime_markers_list_names <- list()
for (i in seq_along(pseudotime_markers_files)) {
	pseudotime_markers_list_names[[i]] <- pseudotime_markers_files[i] %>% 
										str_remove("../analyses/markers/subclustering/RNA_snn_res_10/pseudotime_v2/RNA_snn_res.1/subcluster_dims_10//subclustering_RNA_snn_res_10__") %>% 
										str_remove("__pseudotime_markers.csv")
}
names(pseudotime_markers_list) <- pseudotime_markers_list_names
```

```{r}
length(pseudotime_markers_list_names)
```


```{r}
options(scipen = 1)
```


```{r}
pseudotime_markers_list[[1]]
```


## re-spline




```{r}
pseudotime_cds_files <- list.files(path = "../analyses/rds_files/subclustering/RNA_snn_res_10/pseudotime_v2/RNA_snn_res.1/subcluster_dims_10/", pattern = "_CDSpseudotime.rds", full.names = TRUE)
pseudotime_cds_files
```

```{r}
pseudotime_cds_files[3] %>% 
	str_remove("../analyses/rds_files/subclustering/RNA_snn_res_10/pseudotime_v2/RNA_snn_res.1/subcluster_dims_10//subclustering_RNA_snn_res_10__") %>% 
	str_remove("_CDSpseudotime.rds")
```

```{r}
paste0("_cluster_", sprintf("%03d", clusters_keep))
```


```{r}
pseudotime_cds_files <- data.frame(file_names = pseudotime_cds_files,
		   cluster_number = pseudotime_cds_files %>% 
								str_remove("../analyses/rds_files/subclustering/RNA_snn_res_10/pseudotime_v2/RNA_snn_res.1/subcluster_dims_10//subclustering_RNA_snn_res_10__") %>% 
								str_remove("_CDSpseudotime.rds")
	) %>% 
	filter(cluster_number %in% paste0("cluster_", sprintf("%03d", clusters_keep))) %>% 
	pull(file_names)
pseudotime_cds_files
```



```{r}

smooth_spline_df <- 8
pt_Tfs_matrix_list <- list()

for (i in seq_along(pseudotime_cds_files)) {
    
	message(paste0("Loading .. ", i, " .. of ", length(pseudotime_cds_files)))
    cluster_name <- pseudotime_cds_files[[i]] %>% 
    	str_remove("../analyses/rds_files/subclustering/RNA_snn_res_10/pseudotime_v2/RNA_snn_res.1/subcluster_dims_10//subclustering_RNA_snn_res_10__") %>% 
    	str_remove("_CDSpseudotime.rds")
    temp_cds <- read_rds(file = pseudotime_cds_files[[i]])

    pt_temp_matrix <- exprs(temp_cds)[match(tfs_symbol$gene_symbol,rownames(rowData(temp_cds))),order(pseudotime(temp_cds))]
    # Can also use "normalized_counts" instead of "exprs" to use various normalization methods, for example:
    #     normalized_counts(TFs_cds, norm_method = "log")
    
    pt_temp_matrix <- t(apply(pt_temp_matrix,1,function(x){smooth.spline(x, df = smooth_spline_df)$y}))
    pt_temp_matrix <- t(apply(pt_temp_matrix,1,function(x){(x - mean(x))/sd(x)}))
    rownames(pt_temp_matrix) <- tfs_symbol$gene_symbol
    pt_Tfs_matrix_list[[i]] <- pt_temp_matrix
    
    pt_temp_matrix_df <-  pt_temp_matrix %>% as.data.frame() %>% rownames_to_column("gene")
    write_csv(
        x = pt_temp_matrix_df, 
        file = paste0(
                "../../proj136/analyses/markers/subclustering/RNA_snn_res_10/pseudotime_v2/RNA_snn_res.1/subcluster_dims_10/",
                "subclustering_RNA_snn_res_10__", cluster_name, "__pseudotime_smooth_spline_TFs.csv"
            )
    )
    
    
}

```

```{r}
pt_Tfs_df_list <- list()
for (i in seq_along(pt_Tfs_matrix_list)) {
	message(paste0("Loading .. ", i, " .. of ", length(pt_Tfs_matrix_list)))
    pt_Tfs_df_list[[i]] <-  pt_Tfs_matrix_list[[i]] %>% as.data.frame() %>% rownames_to_column("gene")
}
```





```{r}
length(pt_Tfs_matrix_list)
```

```{r}
pt_Tfs_df_list[[1]]
```


```{r}
reduced_pseudotime_TFs_list <- reduce_dataframes(pt_Tfs_df_list)
```

```{r}
reduced_pseudotime_TFs_list[[1]]
```

```{r}
rownames(reduced_pseudotime_TFs_list[[1]])[1:5]
```

```{r}
names(reduced_pseudotime_TFs_list)[[1]]
```

```{r}
all_expressed_genes <- read_csv("../analyses/all_expressed_genes.csv")
all_expressed_genes
```



```{r}
reduced_all <- list()
for (i in seq_along(reduced_pseudotime_TFs_list)) {
	reduced_all[[i]] <- reduced_pseudotime_TFs_list[[i]]%>%
        filter(!if_all(everything(), ~ .x == "NaN")) %>% 
		rownames_to_column("gene") %>% 
	    filter(gene %in% all_expressed_genes$gene) %>% 
		mutate(subcluster = paste0("subcluster_", i)) %>% 
		dplyr::relocate(subcluster, gene)
}
reduced_all <- bind_rows(reduced_all)
reduced_all
```

```{r}
mean_reduced_tfs <- reduced_all %>% 
	filter(gene %in% repeated_tfs) %>% 
	gather("pseudotime", "smoothed_exp", -subcluster, -gene) %>% 
	mutate(pseudotime = str_remove(pseudotime, "V") %>% as.numeric()) %>% 
	group_by(gene, pseudotime) %>% 
	summarise(mean_exp = mean(smoothed_exp, na.rm = TRUE)) %>% 
	spread(pseudotime, mean_exp) %>% 
	column_to_rownames("gene") %>% 
	select(1:90)
mean_reduced_tfs
```


```{r}
library(RColorBrewer)
library(ComplexHeatmap)
```


```{r}
#K means with 6 groups
htkm <- Heatmap(
  mean_reduced_tfs,
  name                         = "z-score",
  col                          = circlize::colorRamp2(seq(from = -0.5, to = 0.5,length = 11),rev(brewer.pal(11, "Spectral"))),
  show_row_names               = TRUE,
  show_column_names            = FALSE,
  row_names_gp                 = gpar(fontsize = 12),
  km                           = 10,
  row_title_rot                = 0,
  cluster_rows                 = TRUE,
  cluster_row_slices           = FALSE,
  cluster_columns              = FALSE,
  use_raster = FALSE)

#Ward.D2 Hierarchical Clustesubclustered_cluster
hthc <- Heatmap(
  mean_reduced_tfs,
  name                         = "z-score",
  col                          = circlize::colorRamp2(seq(from = -0.5, to = 0.5, length = 11),rev(brewer.pal(11, "Spectral"))),
  show_row_names               = TRUE,
  show_column_names            = FALSE,
  row_names_gp                 = gpar(fontsize = 12),
  clustering_method_rows       = "ward.D2",
  clustering_method_columns    = "ward.D2",
  row_title_rot                = 0,
  cluster_rows                 = TRUE,
  cluster_row_slices           = FALSE,
  cluster_columns              = FALSE,
  use_raster = FALSE)
```



```{r fig.width=8, fig.height=4}
print(htkm)
print(hthc)
```






# Read in rds to extract pseudotime meta data


```{r}
library(monocle3)
```

```{r}
clusters_keep <- c(
	10,16,17,18,19,20,22,23,24,26,27,29,30,33,36,38,39,42,43,46,
	48,49,50,51,53,55,56,57,58,62,63,64,66,69,71,72,73,74,75,81,
	82,84,85,86,87,88,89,90,94,95,98,100,101,105,106,107,109,110,
	113,115,117,121,128,129,132,134,135,138,145,146,148,150,151,
	152,159,161,166,169,171,175,180,182,193,195
)
clusters_keep
```





```{r}
rds_files <- list.files(path = "../analyses/rds_files/subclustering/RNA_snn_res_10/pseudotime_v2/RNA_snn_res.1/subcluster_dims_10/", pattern = "CDSpseudotime.rds", full.names = TRUE)
rds_files
```



```{r}
rds_files <- data.frame(file_names = rds_files,
		   cluster_number = rds_files %>% 
								str_remove("../analyses/rds_files/subclustering/RNA_snn_res_10/pseudotime_v2/RNA_snn_res.1/subcluster_dims_10//subclustering_RNA_snn_res_10__") %>% 
								str_remove("_CDSpseudotime.rds")
	) %>% 
	filter(cluster_number %in% paste0("cluster_", sprintf("%03d", clusters_keep))) %>% 
	pull(file_names)
rds_files
```





```{r}
rds_list <- list()
for (i in seq_along(rds_files)) {
	message(paste0("Loading object .. ", i, " .. of ", length(rds_files)))
	rds_list[[i]] <- read_rds(file = rds_files[[i]])
}
```

```{r}
length(rds_list)
```


```{r}
rds_list_names <- list()
for (i in seq_along(rds_files)) {
	rds_list_names[[i]] <- rds_files[i] %>% 
										str_remove("../analyses/rds_files/subclustering/RNA_snn_res_10/pseudotime_v2/RNA_snn_res.1/subcluster_dims_10//subclustering_RNA_snn_res_10__") %>% 
										str_remove("_CDSpseudotime.rds")
}
names(rds_list) <- rds_list_names
```

```{r}
length(rds_list_names)
```





```{r}

pseudotime_metadata_list <- list()
for (i in seq_along(rds_files)) {
	message(paste0("Extracting pseudotime .. ", i, " .. of ", length(rds_files)))
	pseudotime_metadata_list[[i]] <- data.frame(pseudotime = pseudotime(rds_list[[i]])) %>% 
		rownames_to_column("cell_id") %>% 
		mutate(pseudotime_percent = 100 * pseudotime / max(pseudotime))
}
bind_rows(pseudotime_metadata_list)

```


```{r}
pseudotime_metadata <- FetchData(object = meta_cb, vars = "experiment") %>% 
	rownames_to_column("cell_id") %>% 
	left_join(bind_rows(pseudotime_metadata_list), by = "cell_id")
pseudotime_metadata
```


```{r}
meta_cb <- AddMetaData(object = meta_cb, metadata = pseudotime_metadata$pseudotime, col.name = "pseudotime")
meta_cb <- AddMetaData(object = meta_cb, metadata = pseudotime_metadata$pseudotime_percent, col.name = "pseudotime_percent")

meta_diet <- AddMetaData(object = meta_diet, metadata = pseudotime_metadata$pseudotime, col.name = "pseudotime")
meta_diet <- AddMetaData(object = meta_diet, metadata = pseudotime_metadata$pseudotime_percent, col.name = "pseudotime_percent")
```


```{r fig.height=14, fig.width=14}


genes <- c("pseudotime", "pseudotime_percent")
plist <- FeaturePlot(
    object = meta_cb, 
    reduction = "tsne_240pcs", 
    features = genes,
    combine = FALSE,
    ncol = 1,
    pt.size = 0.3, 
    # min.cutoff = 1,
    # max.cutoff = 3,
    # slot = "counts",
    cols = c("lightsteelblue2", "black"), 
    raster = FALSE,
    raster.dpi = c(1024, 1024),
    order = TRUE,
    coord.fixed = TRUE
)

for (i in seq_along(plist)) {
	plist[[i]]$layers[[1]]$aes_params$alpha = .1
	plot(plist[[i]] + 
		 	scale_colour_continuous_sequential(palette = "Plasma", na.value = "grey90", rev = FALSE) +
    	NoLegend()
    )
}

```




# Save


```{r}
colnames(meta_cb@meta.data)
```



```{r}
colnames(meta_diet@meta.data)
```



```{r}
write_rds(x = meta_cb, file = "../../proj136/analyses/rds_files/New_Meta_cbNeuron_PUBnofacs_noPolIII_v1_alevin_LogNormalize_harmony_integrated_seurat_doublets_removed.rds")
write_rds(x = meta_diet, file = "../../proj136/analyses/rds_files/New_Meta_cbNeuron_PUBnofacs_noPolIII_v1_alevin_LogNormalize_harmony_integrated_seurat_doublets_removed_diet.rds")
```






# Session info

```{r session_info}
sessionInfo()
```



