---
title: "Allen_2025a_Analysis__Figure_4_ALad1_pseudotime"
description: description
author: "Aaron M. Allen"
date: 'Last update: `r date()`'
output:
  html_document:
    number_sections: true
    code_folding: show
    code_download: true
    theme: cerulean
    df_print: paged
    fig_width: 8.5
    fig_height: 5
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: false
    toc_depth: 6
---


```{css, echo = FALSE}
pre, code {white-space:pre !important; overflow-x:auto}
.tocify-item {white-space:pre}
```


```{r echo = FALSE, cache = FALSE}

```


```{r setup, echo = FALSE, cache = FALSE}
ggplot2::theme_set(ggplot2::theme_grey())
ggplot2::theme_update(panel.background = ggplot2::element_rect(fill = "transparent", colour = NA),
                      plot.background = ggplot2::element_rect(fill = "transparent", colour = NA))

rmd_name <- gsub(pattern = ".Rmd", replacement = "", x = knitr::current_input(), ignore.case = TRUE)
knitr::opts_chunk$set(dev = c("png", "cairo_pdf"),
                      dev.args=list(bg="transparent"),
                      fig.align = "center",
                      fig.height = 5,
                      fig.width = 8.5,
                      dpi = 300,
                      pdf.options(encoding = "ISOLatin9.enc"),
                      fig.path = paste0("../analyses/figures/", rmd_name, "-",
                                        format(Sys.time(), "%Y%m%d_%H%M%S"), "/"),
                      fig.retina = 1,
                      warning = TRUE,
                      message = TRUE)
```



# Setup

## Job_ID = `r commandArgs(trailingOnly = TRUE)[2]`
## Input_file = `r commandArgs(trailingOnly = TRUE)[3]`
## Cluster_resolution = `r commandArgs(trailingOnly = TRUE)[4]`
## Subcluster_resolution = `r commandArgs(trailingOnly = TRUE)[5]`
## Subcluster_num_dims = `r commandArgs(trailingOnly = TRUE)[6]`


## Start time

```{r start-time}
set.seed(12021)
today <- format(Sys.time(), "%Y%m%d_%H%M%S")
total_start_time <- Sys.time()
total_start_time
```






## Parameters





## Libraries

```{r packages-etc, warning = FALSE, message = FALSE}

library(Seurat)
library(cowplot)
library(colorspace)
library(tidyverse)

library(monocle3)
library(patchwork)

library(RColorBrewer)
library(ComplexHeatmap)


```




## TFs

### r6.30


```{r load-tfs}
r630 <- read_tsv("../../../aallen/genomes/dmel/r6.30/gtf/dmel-all-r6.30.gtf.gz",
                 col_names = c("sequence","source","feature","start","end","score","strand","phase","attributes")) %>%
                 separate(attributes,c("gene_id","gene_symbol","transcript_id","transcript_symbol"),sep = ";")
# r630

r630_clean <- r630 %>% 
    filter(feature == "gene") %>% 
    select(gene_id, gene_symbol) %>% 
    mutate(gene_id = gene_id %>% 
                       str_remove_all(pattern = '"') %>% 
                       str_remove(pattern = 'gene_id '),
           gene_symbol = gene_symbol %>% 
                           str_remove_all(pattern = '"') %>% 
                           str_remove(pattern = 'gene_symbol ')
           )
# r630_clean

tfs_fbgn <- read_csv("../../../aallen/gene_lists/gene_sets/FlyBase_IDs_allTFs.txt")
# tfs_fbgn

tfs_symbol <- r630_clean %>% 
    filter(gene_id %in% tfs_fbgn$FBgn) %>% 
    select(gene_symbol) %>% 
    mutate(gene_symbol = gene_symbol %>% 
                           str_remove(pattern = ' ')
           )
# tfs_symbol
```





### iso remove

```{r}
iso_genes_remove <- c("fru-plus", "fru-dP1", "fru-dP1-plus", "fruA", "fruB", "fruC",
					  "fruD1", "fruD2", "fruABint", "fruBCint", "fruCint", "fruPr1Ex1",
					  "fruPr1Ex2", "fruCOM", "dsxM", "dsxF", "dsxCOM",
					  "EGFP", "GFP", "mCherry", "Gal4", "GAL4",
					  "lncRNA:roX1", "lncRNA:roX2"
)
```





# Variables

```{r}

set_verbose <- TRUE
num_neighbours <- 60
num_var_genes <- 1000

```


```{r}
input_file <- NULL
cluster_resolution <- "subcluster_clusters"
cluster_dot_resolution <-"subcluster_clusters"

subcluster_resolution <- "RNA_snn_res.1"
subcluster_dot_resolution <- "RNA_snn_res.1"

subcluster_num_dims <- 6
```




# Pseudotime

## Load data


```{r}
ALad1_seu <- read_rds("../../proj136/analyses/rds_files/subclustering_ALad1_metablob_w_LiLabData--round2.rds")
ALad1_seu
```


```{r warning=FALSE, message=FALSE, fig.height=12, fig.width=22}


DimPlot(object = ALad1_seu, 
		group.by = "RNA_snn_res.4",
        reduction = "umap_60pcs", 
		split.by = "technology",
		pt.size = 1,
        label = TRUE, 
        label.box = FALSE) + 
    NoLegend() + 
    coord_fixed()
DimPlot(object = ALad1_seu, 
		group.by = "PN_type_transfer",
        reduction = "umap_60pcs", 
		split.by = "technology",
		pt.size = 1,
        label = TRUE, 
        label.box = FALSE) + 
    NoLegend() + 
    coord_fixed()

genes <- c("Lim3")

plist <- list()
for (i in seq_along(genes)) {
    plist[[i]] <- FeaturePlot(object = ALad1_seu, 
                               features = genes[[i]],
                               split.by = "technology",
                               pt.size = 1, 
                               cols = c("lightsteelblue2", "black"),
                               reduction = "umap_60pcs", 
                               raster = FALSE, 
                               order = TRUE,
                               coord.fixed = TRUE
                              )
}

for (i in seq_along(plist)) {
	# plist[[i]]$layers[[1]]$aes_params$alpha = .1
	plot(plist[[i]])
}


```


```{r}
cell_to_keep <- FetchData(object = ALad1_seu, vars = c("technology", "RNA_snn_res.4")) %>% 
    rownames_to_column("cell_id") %>% 
    filter(technology == "10x") %>% 
    filter(RNA_snn_res.4 != 21) %>% 
    pull(cell_id)
```



```{r}

subclustering_seu <- subset(x = ALad1_seu, cells = cell_to_keep)
subclustering_seu

```







## Find _Imp_ cells



```{r plot-clusters, fig.width=8, fig.height=8}

DimPlot(
		object = subclustering_seu,
		group.by = subcluster_dot_resolution,
		reduction = "umap_60pcs",
		pt.size = 2,
		raster = FALSE,
		label = FALSE,
		label.box = FALSE,
		repel = TRUE) +
    # NoAxes() +
    NoLegend() +
    coord_fixed()

```



```{r fig.height=14, fig.width=14}

genes <- c("Imp", "dati", "br", "pdm3")
plist <- FeaturePlot(
			object = subclustering_seu, 
			reduction = "umap_60pcs", 
			features = genes,
			combine = FALSE,
			ncol = 1,
			pt.size = 2, 
			# min.cutoff = 1,
			# max.cutoff = 2,
			# slot = "counts",
			cols = c("lightsteelblue2", "black"), 
			raster = FALSE,
			raster.dpi = c(1024, 1024),
			order = TRUE,
			coord.fixed = TRUE
		)

for (i in seq_along(plist)) {
	# plist[[i]]$layers[[1]]$aes_params$alpha = .1
	plot(plist[[i]])
}

```


```{r avg-exp}
avg_exp <- AverageExpression(object = subclustering_seu, 
                             assays = "RNA", 
                             slot = "counts",
                             features = c("Imp", "dati", "mamo", "pros", "br", "pdm3"),
                             group.by = subcluster_dot_resolution)

scaled_exp <- avg_exp$RNA %>% 
    as.data.frame() %>% 
    rownames_to_column("gene") %>% 
    gather("cluster", "expression", -gene) %>% 
    spread(gene, expression) %>% 
    column_to_rownames("cluster") %>% 
    as.matrix() %>% 
    scale() %>% 
    as.data.frame() %>% 
    rownames_to_column("cluster") %>% 
    gather("gene", "expression", -cluster) 
# scaled_exp
```


```{r fig.width=10, fig.height=6}
scaled_exp %>% 
    spread(cluster, expression) %>% 
    column_to_rownames("gene") %>% 
    pheatmap::pheatmap(.,
         breaks = seq(-2,2,0.02),
         color = diverging_hcl(201, "Vik"),
         border_color = NA,
         cluster_rows = F,
         cluster_cols = T,
         fontsize = 16,
         angle_col = 90)
```





### All Imp Clusters


```{r}
all_imp_cluster <- scaled_exp %>% 
    filter(gene == "Imp") %>% 
    filter(expression > 0.2) %>% 
    pull(cluster)
all_imp_cluster
```


```{r}
in_conda <- R.home() %>% str_detect("R4_cbrg")
if (!in_conda) {
    all_imp_cluster <- all_imp_cluster %>% str_remove("g")
    all_imp_cluster
}
```



```{r}
Idents(subclustering_seu) <- subcluster_dot_resolution
all_imp_cells <- WhichCells(object = subclustering_seu, idents = all_imp_cluster)
head(all_imp_cells)
```


```{r fig.width=8, fig.height=8}

DimPlot(
		object = subclustering_seu,
		cells.highlight = all_imp_cells,
		sizes.highlight = 2, 
		cols.highlight = "black", 
		cols = "lightsteelblue2",
		reduction = "umap_60pcs",
		pt.size = 2,
		raster = FALSE,
		label = FALSE,
		label.box = FALSE,
		repel = TRUE) +
    # NoAxes() +
    NoLegend() +
    coord_fixed()

```

```{r}
all_imp_metadata <- FetchData(object = subclustering_seu, vars = subcluster_dot_resolution) %>% 
    rownames_to_column("cell_id") %>% 
    mutate(all_imp = if_else(cell_id %in% all_imp_cells, "yes", "no"))

subclustering_seu <- AddMetaData(object = subclustering_seu, metadata = all_imp_metadata$all_imp, col.name = "imp_cells")
```








## Make cds


```{r create-cds}
expression_matrix <- subclustering_seu@assays[["RNA"]]@counts
cell_metadata <- subclustering_seu@meta.data
gene_annotation <-  data.frame(gene = rownames(subclustering_seu)) %>%
    mutate(gene_short_name = gene) %>%
    column_to_rownames("gene")

cds <- new_cell_data_set(expression_matrix,
                         cell_metadata = cell_metadata,
                         gene_metadata = gene_annotation)
```



## Process cds


```{r}

subcluster_resolution <- "PN_type_transfer"
subcluster_dot_resolution <- "PN_type_transfer"

subcluster_num_dims <- 12
num_var_genes <- 2000
num_neighbours <- 20

```


```{r process-object}

var_genes <- VariableFeatures(subclustering_seu)
genes_use <- var_genes[1:min(num_var_genes, length(var_genes))]

cds <- preprocess_cds(cds, method = "PCA", 
                      num_dim = subcluster_num_dims, 
                      norm_method = "log", 
                      use_genes = genes_use, 
                      verbose = set_verbose
                      )

cds <- align_cds(cds, 
                 preprocess_method = "PCA", 
                 alignment_k = num_neighbours, 
                 alignment_group = "experiment",
                 # alignment_group = "orig.ident", 
                 verbose = TRUE
                 )

cds <- reduce_dimension(cds, 
                        max_components = 2,
                        reduction_method = "UMAP",
                        umap.n_neighbors = num_neighbours,
                        cores = 10,
                        verbose = set_verbose
                        )


cds <- cluster_cells(cds, 
                     reduction_method = "UMAP",
                     k = num_neighbours,
                     cluster_method = "louvain",
                     verbose = set_verbose
                     )

cds <- learn_graph(cds,
                   use_partition = FALSE,
                   verbose = set_verbose
                   )

```



## Plot cds


```{r fig.height=14, fig.width=14, message=FALSE, warning=FALSE}


p1 <- plot_cells(cds, 
                 label_groups_by_cluster = FALSE,  
                 color_cells_by = subcluster_dot_resolution,
                 group_label_size = 6,
                 cell_size = 2, 
                 show_trajectory_graph = FALSE) + 
    coord_fixed()
p2 <- plot_cells(cds,
                 label_groups_by_cluster = FALSE,  
                 color_cells_by = "experiment", 
                 group_label_size = 6,
                 cell_size = 2, 
                 show_trajectory_graph = FALSE) + 
    coord_fixed()

dev_genes <- c("Imp","Eip93F","br","dati")
p3 <- plot_cells(cds,
           genes = dev_genes,
           label_cell_groups = FALSE,
           show_trajectory_graph = FALSE,
           cell_size = 1) + 
    coord_fixed()

p4 <- plot_cells(cds,
           color_cells_by = "imp_cells",
           label_cell_groups = FALSE,
           label_leaves = TRUE,
           label_branch_points = TRUE,
           graph_label_size = 1.5,
           cell_size = 2) + 
    coord_fixed()

# plot((p1 + p2)/(p3 + p4))
plot(p1)
plot(p2)
plot(p3)
plot(p4)


```






## Order Cells

```{r root-function}

get_earliest_principal_node <- function(cds, early_born="yes"){
    cell_ids <- which(colData(cds)[, "imp_cells"] == early_born)
    
    closest_vertex <- cds@principal_graph_aux[["UMAP"]]$pr_graph_cell_proj_closest_vertex
    closest_vertex <- as.matrix(closest_vertex[colnames(cds), ])
    root_pr_nodes <- igraph::V(principal_graph(cds)[["UMAP"]])$name[as.numeric(names(which.max(table(closest_vertex[cell_ids,]))))]
    
    root_pr_nodes
}


```


```{r order-cells}

cds <- order_cells(cds, root_pr_nodes=get_earliest_principal_node(cds))

```


```{r fig.width=10, fig.height=10}
plot_cells(cds, 
         label_groups_by_cluster = FALSE,  
         color_cells_by = subcluster_dot_resolution,
         group_label_size = 4,
         cell_size = 2, 
         show_trajectory_graph = FALSE) + 
    coord_fixed()
plot_cells(cds,
           color_cells_by = "pseudotime",
           label_cell_groups = FALSE,
           label_leaves = FALSE,
           label_branch_points = FALSE,
           graph_label_size = 1.5,
           cell_size = 2,
           rasterize = TRUE) +
    coord_fixed()
```





## Coorelate order

```{r}
pseudotime_df <- data.frame(pseudotime = pseudotime(cds)) %>% rownames_to_column("cell_id")
pseudotime_df
```


```{r}
FetchData(object = subclustering_seu, vars = "PN_type_transfer") %>% 
    rownames_to_column("cell_id") %>% 
    left_join(pseudotime_df, by = "cell_id") %>% 
    group_by(PN_type_transfer) %>% 
    summarise(avg_pseudotime = mean(pseudotime)) %>% 
    arrange(avg_pseudotime) %>% 
    filter(str_detect(string = PN_type_transfer, pattern = "u[:digit:][:digit:]", negate = TRUE))
```




```{r}
real_order <- data.frame(rank = 1:11,
                         # cell_type = c("DA4l", "VA6", "DL1", "DC2", "D", "DC3", "VA1d", "VM7v", "VM7d", "VM2", "DM6", "VA1v")
                         cell_type = c("DA4l", "VA6", "DL1", "DC2", "D", "DC3", "VA1d", "VM7", "VM2", "DM6", "VA1v")
                         )
real_order
```


```{r}
recovered_order <- FetchData(object = subclustering_seu, vars = "PN_type_transfer") %>% 
    rownames_to_column("cell_id") %>% 
    left_join(pseudotime_df, by = "cell_id") %>% 
    group_by(PN_type_transfer) %>% 
    summarise(avg_pseudotime = mean(pseudotime)) %>% 
    arrange(avg_pseudotime) %>% 
    filter(str_detect(string = PN_type_transfer, pattern = "u[:digit:][:digit:]", negate = TRUE)) %>% 
    filter(str_detect(string = PN_type_transfer, pattern = "unknown", negate = TRUE)) %>% 
    left_join(real_order, by = c("PN_type_transfer" = "cell_type"))
recovered_order
```



```{r}
kendall_tau <- cor.test(
    real_order$rank, 
    recovered_order$rank,
    method = "kendall")
kendall_tau
```


```{r}
spearman_rank <- cor.test(
    real_order$rank, 
    recovered_order$rank,
    method = "spearman")
spearman_rank
```






## Sig Genes along pseudotime



```{r sig-genes}
modulated_sig_genes <- graph_test(cds, neighbor_graph = "principal_graph", cores = 10)
sig_genes <- row.names(subset(modulated_sig_genes, q_value < 1e-50 & morans_I > 0.2)) %>% dplyr::setdiff(iso_genes_remove) 
sig_genes %>% sort()
```



```{r}
pt.matrix <- exprs(cds)[match(sig_genes,rownames(rowData(cds))),order(pseudotime(cds))]
pt.matrix <- t(apply(pt.matrix,1,function(x){smooth.spline(x, df = 3)$y}))
pt.matrix <- t(apply(pt.matrix,1,function(x){(x - mean(x))/sd(x)}))
rownames(pt.matrix) <- sig_genes
```



```{r}
#K means with 6 groups
htkm <- ComplexHeatmap::Heatmap(
  pt.matrix,
  name                         = "z-score",
  col                          = circlize::colorRamp2(seq(from = -2, to = 2,length = 11),rev(brewer.pal(11, "Spectral"))),
  show_row_names               = TRUE,
  show_column_names            = FALSE,
  row_names_gp                 = gpar(fontsize = 10),
  km                           = 10,
  row_title_rot                = 0,
  cluster_rows                 = TRUE,
  cluster_row_slices           = FALSE,
  cluster_columns              = FALSE,
  use_raster = TRUE)

#Ward.D2 Hierarchical Clustering
hthc <- ComplexHeatmap::Heatmap(
  pt.matrix,
  name                         = "z-score",
  col                          = circlize::colorRamp2(seq(from = -2, to = 2, length = 11),rev(brewer.pal(11, "Spectral"))),
  show_row_names               = TRUE,
  show_column_names            = FALSE,
  row_names_gp                 = gpar(fontsize = 10),
  clustering_method_rows       = "ward.D2",
  clustering_method_columns    = "ward.D2",
  row_title_rot                = 0,
  cluster_rows                 = TRUE,
  cluster_row_slices           = FALSE,
  cluster_columns              = FALSE,
  use_raster = TRUE)
```

```{r fig.width=8, fig.height=20}
print(htkm)
print(hthc)
```





#### sig TFs

```{r}
sig_tf_genes <- dplyr::intersect(sig_genes, tfs_symbol$gene_symbol)
sig_tf_genes
```


```{r}
tf.pt.matrix <- exprs(cds)[match(sig_tf_genes,rownames(rowData(cds))),order(pseudotime(cds))]
tf.pt.matrix <- t(apply(tf.pt.matrix,1,function(x){smooth.spline(x, df = 3)$y}))
tf.pt.matrix <- t(apply(tf.pt.matrix,1,function(x){(x - mean(x))/sd(x)}))
rownames(tf.pt.matrix) <- sig_tf_genes
```


```{r}
#K means with 6 groups
htkm <- Heatmap(
  tf.pt.matrix,
  name                         = "z-score",
  col                          = circlize::colorRamp2(seq(from = -2, to = 2,length = 11),rev(brewer.pal(11, "Spectral"))),
  show_row_names               = TRUE,
  show_column_names            = FALSE,
  row_names_gp                 = gpar(fontsize = 12),
  km                           = 10,
  row_title_rot                = 0,
  cluster_rows                 = TRUE,
  cluster_row_slices           = FALSE,
  cluster_columns              = FALSE,
  use_raster = TRUE)

#Ward.D2 Hierarchical Clustesubclustered_cluster
hthc <- Heatmap(
  tf.pt.matrix,
  name                         = "z-score",
  col                          = circlize::colorRamp2(seq(from = -2, to = 2, length = 11),rev(brewer.pal(11, "Spectral"))),
  show_row_names               = TRUE,
  show_column_names            = FALSE,
  row_names_gp                 = gpar(fontsize = 12),
  clustering_method_rows       = "ward.D2",
  clustering_method_columns    = "ward.D2",
  row_title_rot                = 0,
  cluster_rows                 = TRUE,
  cluster_row_slices           = FALSE,
  cluster_columns              = FALSE,
  use_raster = TRUE)
```



```{r fig.width=8, fig.height=5}
print(htkm)
print(hthc)
```




# Save Data


```{r make-dir}
dir.create(paste0("../analyses/rds_files/subclustering/", cluster_resolution, "/pseudotime/",
  			      subcluster_resolution, "/subcluster_dims_", subcluster_num_dims, "/"), 
		   recursive = TRUE, showWarnings = FALSE)
dir.create(paste0("../analyses/markers/subclustering/", cluster_resolution, "/pseudotime/",
  			      subcluster_resolution, "/subcluster_dims_", subcluster_num_dims, "/"), 
		   recursive = TRUE, showWarnings = FALSE)
```


```{r save-rds}
save_name <- str_remove(string = input_file, pattern = "_integrated.rds")
write_rds(x = subclustering_seu, 
          file = paste0("../analyses/rds_files/subclustering/", cluster_resolution, "/pseudotime/",
          			    subcluster_resolution, "/subcluster_dims_", subcluster_num_dims, "/",
                        save_name, "_CDSpseudotime.rds")
          )

write_csv(x = modulated_sig_genes, 
          file = paste0("../analyses/markers/subclustering/", cluster_resolution, "/pseudotime/",
          			    subcluster_resolution, "/subcluster_dims_", subcluster_num_dims, "/",
                        save_name, "__pseudotime_markers.csv")
          )

pt.matrix.df <-  pt.matrix %>% as.data.frame() %>% rownames_to_column("gene")
write_csv(x = pt.matrix.df, 
          file = paste0("../analyses/markers/subclustering/", cluster_resolution, "/pseudotime/",
          			    subcluster_resolution, "/subcluster_dims_", subcluster_num_dims, "/",
                        save_name, "__pseudotime_smooth_spline_markers.csv")
          )
```





# Session Info

```{r}
sessionInfo()
```





